docs: Debug Menu design document (#893)
platform-platform/monorepo: metrics/web/docs/features/debug_menu/01_debug_menu_design.md

# Debug menu design

## TL;DR

Introducing the `Debug Menu` feature to the Metrics Web Application allows us to access the debug features of the application, such as the `FPS Monitor`. The `Debug Menu` stores configurations locally and uses them to enable or disable some features.

## Hive

> One of the main ideas of the `Debug Menu` feature is to store the configurations locally. That means that we are to use local persistent storage that would provide the application with the current debug configurations if they are enabled and keep these configurations for different sessions.
> For the current implementation, we suggest using the [`Hive` package](https://pub.dev/packages/hive) that uses the [`IndexedDB`](https://developers.google.com/web/ilt/pwa/working-with-indexeddb) to store key-value pairs in [boxes](https://docs.hivedb.dev/#/basics/boxes). Read more about Hive in their [documentation](https://docs.hivedb.dev/#/).
> The local configurations are to be stored using the following document structure that matches one for the [`Firestore Feature Config`](https://github.com/platform-platform/monorepo/blob/master/metrics/web/docs/features/feature_config/01_feature_config_design.md) feature:

Reviewer: ```suggestion
One of the main ideas of the `Debug Menu` feature is to store the configurations locally. That means that we are to use local persistent storage that would provide the application with the current debug configurations if they are enabled and keep these configurations for different sessions.

For the current implementation, we suggest using the [`Hive` package](https://pub.dev/packages/hive) that uses the [`IndexedDB`](https://developers.google.com/web/ilt/pwa/working-with-indexeddb) to store key-value pairs in [boxes](https://docs.hivedb.dev/#/basics/boxes). Read more about Hive in their [documentation](https://docs.hivedb.dev/#/).

The local configurations are to be stored using the following document structure that matches one for the [`Firestore Feature Config`](https://github.com/platform-platform/monorepo/blob/master/metrics/web/docs/features/feature_config/01_feature_config_design.md) feature:
```

The following sequence diagram describes how the application updates the `Local Config` values:

![Update config sequence diagram](http://www.plantuml.com/plantuml/proxy?cache=no&fmt=svg&src=https://github.com/platform-platform/monorepo/raw/debug_menu_document/metrics/web/docs/features/debug_menu/diagrams/debug_menu_update_config_sequence_diagram.puml)

Let's consider the mechanism of updating the `Local Config` values in the application. Assume that the `local_config` box has been opened already using the `OpenLocalConfigStorageUseCase` within the initialize method. The following steps describe how the application acts when the user updates the `Local Config`, e.g. by toggling the FPS Monitor:
1. The UI calls the `toggleFpsMonitor` method of the `DebugMenuNotifier`.
2. The `DebugMenuNotifier` updates the `Local Config` in the `IndexedDB`.
3. The `DebugMenuNotifier` updates the `FpsMonitorLocalConfigViewModel` and notifies the application.

> Now, the user sees the updated UI with the toggled FPS monitor.
> 
> The application does not change the UI until the `local_config` box update succeeds.

Reviewer: ```suggestion
Now, the user sees the updated UI with the toggled FPS monitor. The application does not change the UI until the `local_config` box update succeeds.
```

![Presentation layer diagram](http://www.plantuml.com/plantuml/proxy?cache=no&fmt=svg&src=https://github.com/platform-platform/monorepo/raw/debug_menu_document/metrics/web/docs/features/debug_menu/diagrams/debug_menu_presentation_layer_class_diagram.puml)

The following sequence diagram describes how the application applies `Local Config` values when a user enters the application:

![Read config sequence diagram](http://www.plantuml.com/plantuml/proxy?cache=no&fmt=svg&src=https://github.com/platform-platform/monorepo/raw/debug_menu_document/metrics/web/docs/features/debug_menu/diagrams/debug_menu_read_config_sequence_diagram.puml)

Let's consider the mechanism of applying the `Local Config` values in the application (considering that the `Hive`'s `local_config` box is already open using the `OpenLocalConfigStorageUseCase`). 
When a user enters the application, he or she stays on the `LoadingPage` until the `Local Config` is initialized.

Firstly, the application fetches the `Feature Config` that includes the `isDebugMenuEnabled` configuration value. 
Please consider the following steps performed by the application depending on the `isDebugMenuEnabled` configuration value:

> If the `Debug Menu` feature **is enabled**:

Reviewer: ```suggestion
After the application fetches the `Feature Config` values and ensures the `Debug Menu` feature **is enabled**, the application calls the `initializeLocalConfig` method of the `DebugMenuNotifier`. Then the notifier follows the following steps:
```

    1. The application calls the `initializeLocalConfig` method of the `DebugMenuNotifier`.
    2. The `DebugMenuNotifier` waits until the initialization completes.
    3. The `DebugMenuNotifier` sets the `isLoading` status to `false`.
    4. The application reacts on the `DebugMenuNotifier.isLoading` changes and dismisses the `LoadingPage`. 
    5. The user proceeds to the application.

> If the `Debug Menu` feature **is not enabled**:

Reviewer: ```suggestion
The only difference for the disabled `Debug Menu` feature is that the application calls the `initializeDefaults` method of the `DebugMenuNotifier`. Notifier then uses default values for the debug configurations and doesn't perform any calls to the local storage.
```

> 
>     1. The application calls the `initializeLocalConfigWithDefaults` method of the `DebugMenuNotifier`. 
>     2. The `DebugMenuNotifier` applies the default configuration values set in the constructor.
>     3. The `DebugMenuNotifier` sets the `isLoading` status to `false`. 
>     4. The application reacts on the `DebugMenuNotifier.isLoading` changes and dismisses the `LoadingPage`.
>     5. The user proceeds to the application. 

Reviewer: ```suggestion
```

    
The following sequence diagram describes how the application updates the `Local Config` values:

![Update config sequence diagram](http://www.plantuml.com/plantuml/proxy?cache=no&fmt=svg&src=https://github.com/platform-platform/monorepo/raw/debug_menu_document/metrics/web/docs/features/debug_menu/diagrams/debug_menu_update_config_sequence_diagram.puml)

> Let's consider the mechanism of updating the `Local Config` values in the application (considering that the `Hive`'s `local_config` box is already open using the `OpenLocalConfigStorageUseCase`).

Reviewer: ```suggestion
Let's consider the mechanism of updating the `Local Config` values in the application. Assume that the `local_config` box has been opened already using the `OpenLocalConfigStorageUseCase` within the initialize method. The following steps describe how the application acts when the user updates the `Local Config`, e.g. by toggling the FPS Monitor:
```

> Please consider the following steps performed by the application when the user updates the `Local Config`, e.g. by toggling the FPS Monitor:

Reviewer: ```suggestion
```

    1. The UI calls the `toggleFpsMonitor` method of the `DebugMenuNotifier`.
    2. The `DebugMenuNotifier` updates the `Local Config` in the `IndexedDB`.
    3. The `DebugMenuNotifier` updates the `FpsMonitorLocalConfigViewModel` and notifies the application.
    4. The user sees the updated UI with the toggled FPS monitor.

The application does not change the UI until the `local_config` box update succeeds.

> When the user exits the application, the application calls the `dispose` method of the `DebugMenuNotifier`. The `dispose` method, in its turn, calls the `CloseLocalConfigStorage` usecase, that closes the `local_config` box.

Reviewer: ```suggestion
The `DebugMenuNotifier` is implemented to close all the opened boxes when the notifier is disposed (not in the widget tree). Currently, the notifier calls the `CloseLocalConfigStorage` usecase when the user exits the application.
```
